cpp/src/community/detail/mis_impl.cuh

```diff
@@ -0,0 +1,729 @@
+                       [] __device__(auto flags) {
+                         return (thrust::get<0>(flags) == 0) && (thrust::get<1>(flags) == 0) &&
+                                (thrust::get<2>(flags) > 0);
+                       });
+
+    if (multi_gpu) {
+      nr_remaining_vertices_to_check = host_scalar_allreduce(handle.get_comms(),
+                                                             nr_remaining_vertices_to_check,
+                                                             raft::comms::op_t::SUM,
+                                                             handle.get_stream());
+    }
+
+    vertex_t nr_discared_and_included =
+      thrust::count_if(handle.get_thrust_policy(),
+                       thrust::make_zip_iterator(thrust::make_tuple(
+                         mis_inclusion_flags.begin(), discard_flags.begin(), out_degrees.begin())),
+                       thrust::make_zip_iterator(thrust::make_tuple(
+                         mis_inclusion_flags.end(), discard_flags.end(), out_degrees.end())),
+                       [] __device__(auto flags) {
+                         return (thrust::get<0>(flags) > 0) && (thrust::get<1>(flags) > 0) &&
+                                (thrust::get<2>(flags) > 0);
+                       });
+
+    if (multi_gpu) {
+      nr_discared_and_included = host_scalar_allreduce(
+        handle.get_comms(), nr_discared_and_included, raft::comms::op_t::SUM, handle.get_stream());
+    }
+    cudaDeviceSynchronize();
+
+    std::cout << " number_of_vertices:       " << number_of_vertices << std::endl;
+    std::cout << " nr_discared_and_included: " << nr_discared_and_included << std::endl;
+    std::cout << " included : discarded:     " << nr_include_vertices << ":"
+              << nr_discarded_vertices << std::endl;
+    std::cout << " included + discarded:     " << (nr_include_vertices + nr_discarded_vertices)
+              << std::endl;
+
+    std::cout << "remaining_vts_to_check:   " << nr_remaining_vertices_to_check << std::endl;
+
+    if (nr_remaining_vertices_to_check == 0) { break; }
+  }
+
+  vertex_t nr_vertices_in_mis = thrust::count_if(handle.get_thrust_policy(),
+                                                 mis_inclusion_flags.begin(),
+                                                 mis_inclusion_flags.end(),
+                                                 [] __device__(auto flag) { return flag > 0; });
+
+  thrust::copy_if(handle.get_thrust_policy(),
+                  vertex_begin,
+                  vertex_end,
+                  mis_inclusion_flags.begin(),
+                  mis.begin(),
+                  [] __device__(auto flag) { return flag > 0; });
+
+  mis.resize(nr_vertices_in_mis, handle.get_stream());
+
+  // if (debug) {
+  cudaDeviceSynchronize();
+  std::cout << "Found mis of size " << mis.size() << std::endl;
+  cudaDeviceSynchronize();
+  // raft::print_device_vector("mis", mis.data(), mis.size(), std::cout);
+  // }
+
+  return mis;
```

Comment:
And I think this code is overly complicated with unnecessarily many temporary state variables.

Can't you just do the following?
```
1. compute out degrees.
2. initialize remaining_vertices to all non-zero out degree vertices.
3. initialize ranks to out-weight-sums.
4. while (remaining_vertices.size() > 0)
4-1. Pick 50% of remaining_vertices as candidates.
4-2. Call per_v_transform_reduce_incoming_e and per_v_transform_reduce_outgoing_e to find the maximum rank neighbor.
If the maximum neighbor rank is std::numeric_limits<weight_t>::max(), remove from `remaining_vertices`.
If the maximum neighbor rank is lower than the vertex's own rank, remove from `remaining_vertices` and add to `mis`. Also, set the rank to std::numeric_limits<<weight_t>::max().
5. return `mis`. 
```

Reply:
And you may create a temporary copy of ranks and set ranks to std::numeric_limits<weight_t>::lowest() if it's not already std::numeric_limits<weight_t>::max() (included in the MIS) and if the vertex is not a candidate.

Alternatively, you may maintain a flag vector indicating whether a vertex is a candidate or not.